pkg/server/home_workspaces.go

```diff
@@ -0,0 +1,661 @@
+			return []string{cluster.Join(o.Name).String()}, nil
+		},
+	})
+}
+
+type searchFunc func(workspace logicalcluster.Name, name string) (bool, error)
+type mockableFunctions struct {
+	searchHomeWorkspaceRBACLocally func(homeWorkspace logicalcluster.Name, userName string) (found bool, err error)
+	createHomeRBAC                 func(ctx context.Context, userName string, homeWorkspace logicalcluster.Name) error
+	checkWorkspaceLocally          func(workspaceName logicalcluster.Name, isHome bool, userName string) (found, retryLater bool, retryAfterSeconds int, checkError error)
+	tryToCreate                    func(ctx context.Context, userName string, workspaceToCheck logicalcluster.Name, workspaceType tenancyv1alpha1.ClusterWorkspaceTypeName) (retryLater bool, retryAfterSeconds int, createError error)
+}
+
+type homeWorkspaceHandler struct {
+	apiHandler http.Handler
+
+	authz authorizer.Authorizer
+
+	createClusterRole        func(ctx context.Context, workspace logicalcluster.Name, cr *rbacv1.ClusterRole) error
+	createClusterRoleBinding func(ctx context.Context, workspace logicalcluster.Name, crb *rbacv1.ClusterRoleBinding) error
+
+	createClusterWorkspace func(ctx context.Context, workspace logicalcluster.Name, cw *tenancyv1alpha1.ClusterWorkspace) error
+	getClusterWorkspace    func(ctx context.Context, workspace logicalcluster.Name, name string) (*tenancyv1alpha1.ClusterWorkspace, error)
+
+	getLocalClusterWorkspace      func(fullName logicalcluster.Name) (*tenancyv1alpha1.ClusterWorkspace, error)
+	searchLocalClusterRole        searchFunc
+	searchLocalClusterRoleBinding searchFunc
+	synced                        func() bool
+
+	homePrefix               logicalcluster.Name
+	bucketLevels, bucketSize int
+	creationDelay            int
+
+	// This is only for testing
+	mocks mockableFunctions
+}
+
+func (h *homeWorkspaceHandler) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	if !h.synced() {
+		h.apiHandler.ServeHTTP(rw, req)
+		return
+	}
+
+	ctx := req.Context()
+	var workspaceName logicalcluster.Name
+	if lcluster, err := request.ValidClusterFrom(ctx); err != nil {
+		responsewriters.InternalError(rw, req, err)
+		return
+	} else {
+		workspaceName = lcluster.Name
+	}
+	user, ok := request.UserFrom(ctx)
+	if !ok {
+		err := errors.New("No user in HomeWorkspaces filter !")
+		responsewriters.InternalError(rw, req, err)
+		return
+	}
+	requestInfo, ok := request.RequestInfoFrom(ctx)
+	if !ok {
+		responsewriters.InternalError(rw, req, errors.New("No request Info"))
+		return
+	}
+
+	if workspaceName == tenancyv1alpha1.RootCluster &&
+		requestInfo.IsResourceRequest &&
+		requestInfo.Verb == "get" &&
+		requestInfo.APIGroup == tenancyv1beta1.SchemeGroupVersion.Group &&
+		requestInfo.Resource == "workspaces" &&
+		requestInfo.Name == "~" {
+		var homeWorkspace *tenancyv1beta1.Workspace
+		homeWorkspaceFullName := h.getHomeWorkspaceName(user.GetName())
+		homeClusterWorkspace, err := h.getLocalClusterWorkspace(homeWorkspaceFullName)
```

Comment:
fullName and local name are no known concepts. What are these? home `logicalcluster.Name` and `ClusterWorkspace.Name`?

Reply:
I renamed those names to remove the confusion in commit 1f37994d81146bb38fbecd97319d1a5a24eba461